src/main/combatLogParser.ts

```diff
@@ -0,0 +1,434 @@
+                if (char === '"') {
+                    inQuotedString = false;
+                    continue;
+                }
+
+            } else {
+                switch (char) {
+                case ',':
+                    if (openListCount > 0) {
+                        listItems.at(-1)?.push(value);
+                    } else {
+                        this.addArg(value);
+                    }
+
+                    value = '';
+                    break;
+
+                case '"':
+                    inQuotedString = true;
+                    break;
+
+                case '[':
+                case '(':
+                    listItems.push([]);
+                    openListCount++;
+                    break;
+
+                case ']':
+                case ')':
+                    if (!listItems.length) {
+                        throw `Unexpected ${char}. No list is open.`;
+                    }
+
+                    if (value) {
+                        listItems.at(-1)?.push(value);
+                    }
+
+                    value = listItems.pop();
+                    openListCount--;
+                    break;
+                }
+
+                continue;
+            }
+
+            value += char;
+        }
+
+        if (value) {
+            this.addArg(value);
+        }
+
+        if (openListCount > 0) {
+            throw `Unexpected EOL. There are ${openListCount} open list(s).`
+        }
+    }
+
+    /**
+     * Add an argument to the list
+     */
+    private addArg(value: any): void {
+        this._args.push(value);
+        this._argsListLen = this._args.length;
+    }
+
+    toString(): string {
+        return this.original;
+    }
+}
+
+/**
+ * Combat log parser and monitoring class.
+ *
+ * Watches one or more directories for combat logs, read the new data from them
+ * and emits events to act on the combat log events elsewhere.
+ */
+class CombatLogParser extends EventEmitter {
+    private _options: CombatLogParserOptionsType;
+    private _handlers: { [key: string]: CombatLogMonitorHandlerType } = {};
+
+    /**
+     * If set, any handler receiving data which _ISN'T_ the one
+     * given here will be ignored. This is to avoid multiple log files receiving
+     * data at the same time which shouldn't happen, but can.
+     */
+    private _handlerLock?: CombatLogMonitorHandlerType;
+    private _handlerLockTimeout: any;
+
+    private readonly _tailOptions = {
+        flushAtEOF: true
+    };
+
+    constructor(options: CombatLogParserOptionsType) {
+        super();
+
+        this._options = options;
+    };
+
+    /**
+     * Start watching a path, if it's a valid WoW logs directory.
+     * This is checked via `getWowFlavour()` which looks for the file '../.flavour.info'
+     * relative to the log directory, which all flavours of WoW have.
+     */
+    watchPath(pathSpec: string): void {
+        pathSpec = path.resolve(pathSpec);
+        if (pathSpec in this._handlers) {
```

Comment:
Why do this and not simply return if we are already watching it?

Reply:
That's a damn good question to which I don't have a useful answer. 🤷‍♂️